### PR TITLE
New ThreadPlanSingleThreadTimeout to resolve potential deadlock in single thread stepping

### PR DESCRIPTION
**Pull request by jeffreytan81**
This PR introduces a new `ThreadPlanSingleThreadTimeout` that will be used to address potential deadlock during single-thread stepping. 

While debugging a target with a non-trivial number of threads (around 5000 threads in one example target), we noticed that a simple step over can take as long as 10 seconds. Enabling single-thread stepping mode significantly reduces the stepping time to around 3 seconds. However, this can introduce deadlock if we try to step over a method that depends on other threads to release a lock. 

To address this issue, we introduce a new `ThreadPlanSingleThreadTimeout` that can be controlled by the `target.process.thread.single-thread-plan-timeout` setting during single-thread stepping mode. The concept involves counting the elapsed time since the last internal stop to detect overall stepping progress. Once a timeout occurs, we assume the target is not making progress due to a potential deadlock, as mentioned above. We then send a new async interrupt, resume all threads, and `ThreadPlanSingleThreadTimeout` completes its task. 

To support this design, the major changes made in this PR are:
1. `ThreadPlanSingleThreadTimeout` is popped during every internal stop and reset (re-pushed) to the top of the stack (as a leaf node) during resume. This is achieved by always returning `true` from `ThreadPlanSingleThreadTimeout::DoPlanExplainsStop()` and `ThreadPlanSingleThreadTimeout::MischiefManaged()`. 
2. A new thread-specific async interrupt stop is introduced, which can be detected/consumed by `ThreadPlanSingleThreadTimeout`. 
3. The clearing of branch breakpoints in the range thread plan has been moved from `DoPlanExplainsStop()` to `ShouldStop()`, as it is not guaranteed that it will be called. 

The detailed design is discussed in the RFC below:
[https://discourse.llvm.org/t/improve-single-thread-stepping/74599](https://discourse.llvm.org/t/improve-single-thread-stepping/74599)

### CODE REVIEW

**Comment by jeffreytan81** [comment]
@jimingham, friendly ping. @clayborg mentioned that you are the sole domain expert on ThreadPlan. Can you help to review this change? Thanks

**Comment by jimingham** [comment]
Thanks for working up this patch.  I should be able to get time to look at this next week.

File: lldb/include/lldb/Target/Process.h [hunk 1]
```diff
@@ -1312,11 +1312,13 @@ class Process : public std::enable_shared_from_this<Process>,
...
   size_t GetThreadStatus(Stream &ostrm, bool only_threads_with_stop_reason,
                          uint32_t start_frame, uint32_t num_frames,
-                         uint32_t num_frames_with_source,
-                         bool stop_format);
+                         uint32_t num_frames_with_source, bool stop_format);
 
   void SendAsyncInterrupt();
 
+  // Send an async interrupt and receive stop from a specific /p thread.
+  void SendAsyncInterrupt(Thread *thread);
```
Comment (jimingham) [inline lldb/include/lldb/Target/Process.h:1320]: Do we need two API's here?  Why not have thread default to nullptr and document that that means we let the system choose which thread gets the interrupt?  It's also worth noting here that the "receive stop from specific thread" is artificial, so you might say something like "and attribute the stop to a specific thread" instead to make it clear we don't require the ability to actually have that thread receive the signal.

File: lldb/include/lldb/Target/ThreadPlan.h [hunk 2]
```diff
@@ -590,6 +591,7 @@ class ThreadPlanNull : public ThreadPlan {
   const Status &GetStatus() { return m_status; }
 
 protected:
+  friend class ThreadPlanSingleThreadTimeout;
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlan.h:594]: What is this needed for?  It's a little odd having a derived class friended to the base class, since you could make whatever the derived class needed protected rather than private.

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+#include "lldb/Utility/State.h"
+
+#include <thread>
+
+namespace lldb_private {
+
+//
+// Thread plan used by single thread execution to issue timeout. This is useful
+// to detect potential deadlock in single thread execution. The timeout measures
+// the elapsed time from the last internal stop and got reset by each internal
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:25]: "gets reset by each internal stop"

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+// stops to ensure we are accurately detecting execution not moving forward.
+// This means this thread plan  can be created/destroyed multiple times by the
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:27]: "may" is better than "can" and there's one too many spaces.

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  static bool IsAlive();
+
+  enum class State {
+    WaitTimeout,    // Waiting for timeout.
+    AsyncInterrupt, // Async interrupt has been issued.
+    Done,           // Finished resume all threads.
+  };
+
+  static std::mutex s_mutex;
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:70]: I'll have to see how you use this as I go along, but having statics here worries me.  How are you going to support the case where lldb has two targets, each of which are running using this timeout feature?

Comment (jeffreytan81) [reply lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:70]: This is a good point. I was storing some static states so that we can ensure: 
1. Singleton (only one ThreadPlanSingleThreadTimeout is alive/used anytime). 
2. If there is ever a race that, the `ThreadPlanSingleThreadTimeout` is in `State::AsyncInterrupt` but it got popped by another internal stop before async interrupt received. I want to remember the previous/last state, and the newly created `ThreadPlanSingleThreadTimeout` can be re-created from previous state.

Instead, I will look into storing these static states into parent controlling ThreadPlan (ThreadPlanStepOverRange in this case).

Comment (jimingham) [reply lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:70]: That would be better, but if you find you do need a singleton, you could also store that either in the Thread or the Process, depending on the scope at which you need them.

File: lldb/include/lldb/Target/ThreadPlan.h (hunk 2 above)
Comment (jeffreytan81) [reply lldb/include/lldb/Target/ThreadPlan.h]: Hmm, I thought there is a reason I can't remember, apparently not. Removed. Thanks.

File: lldb/include/lldb/Target/Process.h (hunk 1 above)
Comment (jeffreytan81) [reply lldb/include/lldb/Target/Process.h]: I chose overloading because I remember reading some guideline to prefer overloading than default parameter. Google finds one:
https://stackoverflow.com/questions/15876590/overload-a-method-or-use-default-values-c

But I do not care that much really, so changed and made the comment more clear.

**Comment by jeffreytan81** [comment]
@jimingham, I have revised the PR, let me know when you can review again. Thanks.

File: lldb/include/lldb/Target/Process.h [hunk 6]
```diff
@@ -1312,10 +1312,16 @@ class Process : public std::enable_shared_from_this<Process>,
...
   size_t GetThreadStatus(Stream &ostrm, bool only_threads_with_stop_reason,
                          uint32_t start_frame, uint32_t num_frames,
-                         uint32_t num_frames_with_source,
-                         bool stop_format);
+                         uint32_t num_frames_with_source, bool stop_format);
 
-  void SendAsyncInterrupt();
+  /// Send an async interrupt request.
+  ///
+  /// If \a thread is specified the async interrupt stop will be attributed the
```
Comment (jimingham) [inline lldb/include/lldb/Target/Process.h:1319]: "will be attributed to the" - missing "to"

File: lldb/include/lldb/Target/Process.h [hunk 7]
```diff
@@ -1312,10 +1312,16 @@ class Process : public std::enable_shared_from_this<Process>,
...
+  /// specified thread.
+  ///
+  /// \param[in] thread
+  ///     The thread from which to attribute the async interrupt stop to.
```
Comment (jimingham) [inline lldb/include/lldb/Target/Process.h:1323]: "The thread the async interrupt will be attributed to"

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+public:
+  struct TimeoutInfo {
+    ThreadPlanSingleThreadTimeout *m_instance = nullptr;
+    ThreadPlanSingleThreadTimeout::State m_last_state = State::WaitTimeout;
+  };
+
+  ~ThreadPlanSingleThreadTimeout() override;
+
+  // Create a new instance from fresh new state.
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:49]: Since this API doesn't return the Created thread plan, then this API has to do more than just create it - it also pushes the plan, and sometimes doesn't.  So CreateNew is not a good name, and the description of what the function does is too abbreviated.

File: lldb/include/lldb/Target/ThreadPlanStepOverRange.h [hunk 9]
```diff
@@ -42,8 +45,13 @@ class ThreadPlanStepOverRange : public ThreadPlanStepRange,
 
   void SetupAvoidNoDebug(LazyBool step_out_avoids_code_without_debug_info);
   bool IsEquivalentContext(const SymbolContext &context);
+  // Clear and create a new ThreadPlanSingleThreadTimeout to detect if program
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanStepOverRange.h:48]: ThreadPlanStepOverRange isn't the only thread plan that we want to have this behavior for.  For instance, it would be great to do this for ThreadPlanStepOut, which we don't even try to run on one thread, IIRC, and in order not to duplicate code, we will want to switch the ThreadPlanCallUserExpression over to this method as well.  So the code to handle the interrupt has to be shareable - maybe a mix-in class that does the management of the interrupt?

File: lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.h [hunk 10]
```diff
@@ -440,6 +440,12 @@ class ProcessGDBRemote : public Process,
   void HandleStopReply() override;
   void HandleAsyncStructuredDataPacket(llvm::StringRef data) override;
 
+  // Handle thread specific async interrupt and return the original thread
+  // that requested the async interrupt. It can be null if original thread
+  // has exited.
+  lldb::ThreadSP HandleThreadAsyncInterrupt(uint8_t signo,
```
Comment (jimingham) [inline lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.h:446]: I can't see any reason why the ability to manage this interrupt event is GDBRemote specific.  So this should be a virtual method in Process, and overridden here.

File: lldb/source/Target/Process.cpp [hunk 11]
```diff
@@ -3905,9 +3911,14 @@ thread_result_t Process::RunPrivateStateThread(bool is_secondary_thread) {
 
       if (interrupt_requested) {
         if (StateIsStoppedState(internal_state, true)) {
-          // We requested the interrupt, so mark this as such in the stop event
-          // so clients can tell an interrupted process from a natural stop
-          ProcessEventData::SetInterruptedInEvent(event_sp.get(), true);
+          // Oly mark interrupt event if it is not thread specific async
```
Comment (jimingham) [inline lldb/source/Target/Process.cpp:3914]: Only

**Comment by jeffreytan81** [comment]
@jimingham , let me know any other comments/thoughts you have.

**Comment by jimingham** [comment]
I will get to this, WWDC is close upon us so we're all rather busy at present.

Jim

**Comment by jeffreytan81** [comment]
@jimingham, hope WWDC is going well. Do you have time to review this now? Thanks.

File: lldb/include/lldb/Target/Process.h [hunk 12]
```diff
@@ -2838,6 +2844,14 @@ void PruneThreadPlans();
     return std::nullopt;
   }
 
+  /// Handle thread specific async interrupt and return the original thread
+  /// that requested the async interrupt. It can be null if original thread
+  /// has exited.
```
Comment (jimingham) [inline lldb/include/lldb/Target/Process.h:2849]: You should say what the `description` string is for.

File: lldb/include/lldb/Target/TimeoutResumeAll.h [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+//===----------------------------------------------------------------------===//
+
+#ifndef LLDB_TARGET_TIMEOUTRESUMEALL_H
+#define LLDB_TARGET_TIMEOUTRESUMEALL_H
+
+#include "lldb/Target/ThreadPlanSingleThreadTimeout.h"
+
+namespace lldb_private {
+
+// Mixin class that provides capability for ThreadPlan that supports single
```
Comment (jimingham) [inline lldb/include/lldb/Target/TimeoutResumeAll.h:16]: Grammar:

// Mixin class that provides the capability for ThreadPlan to support single
// thread execution that resumes all threads after a timeout.

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+//
+// Thread plan used by single thread execution to issue timeout. This is useful
+// to detect potential deadlock in single thread execution. The timeout measures
+// the elapsed time from the last internal stop and gets reset by each internal
+// stop to ensure we are accurately detecting execution not moving forward.
+// This means this thread plan may be created/destroyed multiple times by the
+// parent execution plan.
+//
+// When timeout happens, the thread plan resolves the potential deadlock by
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:30]: "When a timeout happens"

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+// issuing a thread specific async interrupt to enter stop state, then all
+// threads execution are resumed to resolve the potential deadlock.
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:32]: "then execution is resumed with all threads running to resolve the potential deadlock"

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 16]
```diff
@@ -0,0 +1,102 @@
...
+//
+class ThreadPlanSingleThreadTimeout : public ThreadPlan {
+  enum class State {
+    WaitTimeout,    // Waiting for timeout.
+    AsyncInterrupt, // Async interrupt has been issued.
+    Done,           // Finished resume all threads.
+  };
+
+public:
+  struct TimeoutInfo {
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:42]: This isn't something you need to do in this patch, but it would be good to allow different thread plan invocations to use different timeouts, and not just depend on the thread timeout.  For instance, we should be able to reimplement RunThreadPlan with this infrastructure.  But the timeouts you want for expression evaluation and stepping are likely to be very different (and different executions also use different timeouts).  So being able to pass in the timeout, rather than relying on one thread specific one will be necessary.  We could for instance add the timeout to the TimeoutInfo and plumb that so that the ThreadPlan that mixes in TimeoutResumeAll can pass in a timeout.

File: lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    ThreadPlanSingleThreadTimeout *m_instance = nullptr;
+    ThreadPlanSingleThreadTimeout::State m_last_state = State::WaitTimeout;
+  };
+
+  ~ThreadPlanSingleThreadTimeout() override;
+
+  // If input \param thread is running in single thread mode, push a
+  // new ThreadPlanSingleThreadTimeout based on timeout setting from fresh new
+  // state. The reference of \param info is passed in so that when
+  // ThreadPlanSingleThreadTimeout got popped out its last state can be stored
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanSingleThreadTimeout.h:52]: "popped out" is confusing.  We just say "popped" for the most part.

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+                 thread, eVoteNo, eVoteNoOpinion),
+      m_info(info), m_state(State::WaitTimeout), m_exit_flag(false) {
+  m_timer_thread = std::thread(TimeoutThreadFunc, this);
+  m_info.m_instance = this;
+  m_state = m_info.m_last_state;
+}
+
+ThreadPlanSingleThreadTimeout::~ThreadPlanSingleThreadTimeout() {
+  m_info.m_instance = nullptr;
+  if (m_state == State::Done)
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:39]: This seems like an odd thing to do in the destructor.  Why is this necessary?

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+    m_state = State::WaitTimeout;
+}
+
+void ThreadPlanSingleThreadTimeout::GetDescription(
+    Stream *s, lldb::DescriptionLevel level) {
+  s->Printf("Single thread timeout, state(%s)", StateToString(m_state).c_str());
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:45]: I think it would be super handy to print the time remaining here.

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+  // Do not create timeout if we are not stopping other threads.
+  if (!thread.GetCurrentPlan()->StopOthers())
+    return;
+
+  auto timeout_plan = new ThreadPlanSingleThreadTimeout(thread, info);
+  ThreadPlanSP thread_plan_sp(timeout_plan);
+  auto status = thread.QueueThreadPlan(thread_plan_sp,
+                                       /*abort_other_plans*/ false);
+  Log *log = GetLog(LLDBLog::Step);
+  LLDB_LOGF(log, "ThreadPlanSingleThreadTimeout pushing a brand new one");
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:74]: definitely want the timeout in this log message.

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 21]
```diff
@@ -0,0 +1,217 @@
...
+  // Do not create timeout if we are not stopping other threads.
+  if (!thread.GetCurrentPlan()->StopOthers())
+    return;
+
+  auto timeout_plan = new ThreadPlanSingleThreadTimeout(thread, info);
+  ThreadPlanSP thread_plan_sp(timeout_plan);
+  auto status = thread.QueueThreadPlan(thread_plan_sp,
+                                       /*abort_other_plans*/ false);
+  Log *log = GetLog(LLDBLog::Step);
+  LLDB_LOGF(log, "ThreadPlanSingleThreadTimeout reset from previous state");
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:95]: Also here, for debugging purposes it will be really handy to have the timeout value here.

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 22]
```diff
@@ -0,0 +1,217 @@
...
+  m_wakeup_cv.notify_one();
+  // Wait for timer thread to exit.
+  m_timer_thread.join();
+}
+
+bool ThreadPlanSingleThreadTimeout::DoPlanExplainsStop(Event *event_ptr) {
+  lldb::StateType stop_state =
+      Process::ProcessEventData::GetStateFromEvent(event_ptr);
+  Log *log = GetLog(LLDBLog::Step);
+  LLDB_LOGF(
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:125]: This is another place where it will probably be useful to see the timeout value if we're logging.

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 23]
```diff
@@ -0,0 +1,217 @@
...
+      StateAsCString(stop_state));
+  return true;
+}
+
+lldb::StateType ThreadPlanSingleThreadTimeout::GetPlanRunState() {
+  return GetPreviousPlan()->GetPlanRunState();
+}
+
+void ThreadPlanSingleThreadTimeout::TimeoutThreadFunc(
+    ThreadPlanSingleThreadTimeout *self) {
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:137]: Depending on how much this gets used, we might not want to spin up this thread afresh every time we invoke a single-thread timeout style thread plan.  There will only ever be one client of this timeout at a time, since this is to implement running only one thread, and gets cleared each time you stop.  So it would be more efficient to have a worker thread you spin up the first time someone does a one-thread-with-timeout and have it just be a timeout-server thread.
I think it's likely we will want to do that, but that can be for a follow-up patch.

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 24]
```diff
@@ -0,0 +1,217 @@
...
+  std::unique_lock<std::mutex> lock(self->m_mutex);
+  uint64_t timeout_in_ms = self->GetThread().GetSingleThreadPlanTimeout();
+  self->m_wakeup_cv.wait_for(lock, std::chrono::milliseconds(timeout_in_ms),
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:140]: It would be good to log both before and after the wait call here.  I think that will help read the step logs.

File: lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp [hunk 25]
```diff
@@ -0,0 +1,217 @@
...
+  LLDB_LOGF(log, "ThreadPlanSingleThreadTimeout::MischiefManaged() called.");
+  // Need to reset timer on each internal stop/execution progress.
+  return true;
+}
+
+bool ThreadPlanSingleThreadTimeout::ShouldStop(Event *event_ptr) {
+  return HandleEvent(event_ptr);
+}
+
+void ThreadPlanSingleThreadTimeout::SetStopOthers(bool new_value) {
```
Comment (jimingham) [inline lldb/source/Target/ThreadPlanSingleThreadTimeout.cpp:165]: This is assuming that the SingleThreadTimeout plan is always going to be pushed on behalf of the plan directly above it.  That seems like a valid assumption to me, but it's worth calling out somewhere.

File: lldb/include/lldb/Target/ThreadPlanStepRange.h [hunk 26]
```diff
@@ -58,8 +58,13 @@ class ThreadPlanStepRange : public ThreadPlan {
   // run' plan, then just single step.
   bool SetNextBranchBreakpoint();
 
+  // Whether the input stop info is caused by the next branch breakpoint.
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlanStepRange.h:61]: This routine does NOT distinguish between the cases where the next branch breakpoint is the ONLY breakpoint at this site, and where it's one of many breakpoints that share this site.  Later, you distinguish between the two cases where this is used (e.g. in ThreadPlanStepRange::NextRangeBreakpointExplainsStop.
It might be interesting to see if it makes sense to incorporate the logic that does that into IsNextBranchBreakpointStop.  If it does that would allow some code reuse.  If it does not, you should note in the comment here that it does not distinguish between the two cases, so someone doesn't make a mistake later on.

**Comment by jimingham** [comment]
I left some more comments, most of them either nits or future work - unless you want to do them here...

This makes sense to me, so it's likely right.  But the tests you've written so far are just about making sure that deadlocks are resolved in a simple stepping situation where nothing user-visible happens while you are stepping.

We need to make sure that if the "next" is interrupted - for instance by hitting a breakpoint in the code being stepped over, that once stopped, lldb can do other step operations having stopped and they work as expected.  

Also, in lldb, if you do a `next` and that steps over a function that hits a breakpoint, the next stays on the stack and then "continue" will resume execution of the "next" ThreadPlan, completing the step.  We should make sure that works and keeps tracking the single thread in a case where there's no deadlock to resolve, or goes on to resolve the deadlock.

We also really want to make sure that the async interrupt that you use here and the one that the user uses (through SBProcess.SendAsyncInterrupt) don't interfere with one another.  So step over something with a really long user step timeout, then send an asynchronous interrupt and make sure we stop with the right stop reason.  And then make sure the next continues successfully after that.

Also, if we can it would be nice to test that this succeeds in NOT allowing the other threads to run before the timeout elapses.  You could have your worker thread stop right before code that's going to crash, and then switch to the main thread and do a bunch of steps over code that calls some function that will do a pthread yield.  If the process doesn't crash after those steps, you're good.

**Comment by jimingham** [comment]
> On Jul 8, 2024, at 12:55 PM, jeffreytan81 ***@***.***> wrote:
> 
> 
> @jimingham <https://github.com/jimingham>, thanks for the comment.
> 
> When I try to add a new testcase for stepping over a function with a user set breakpoint, there is a bug found -- the user set breakpoint did not stop with ThreadPlanSingleThreadTimeout present (which is c case in your RFC comment)
> 
> I did some investigation, when we issue normal step-over, the all thread plans in the stack return false for PlanExplainsStop, so base thread plan got a chance to check thread's stop info and stop the execution.
> 
> However, while doing single thread stepping, ThreadPlanSingleThreadTimeout as leaf plan always return true for PlanExplainsStop, causing the logic to use the second while loop in Thread::ShouldStop(), eventually, ThreadPlanStepOut::MischiefManaged() returns false causing the loop to terminate early without giving base thread plan to decide should stop or not:
> https://github.com/llvm/llvm-project/blob/393eff4e02e7ab3d234d246a8d6912c8e745e6f9/lldb/source/Target/Thread.cpp#L905-L907
> 
> So overall the Thread::ShouldStop() returns false and ignores the hit inner breakpoint during stepping. Any suggestion how to fix this?
> 

Why does the ThreadPlanSingleThreadTimeout have to claim to explain stops (like a random breakpoint hit) that on the face of it it's not responsible for?  What would happen if it just didn't explain this breakpoint, and instead uses marking itself as stale when a user stop like this comes in to make sure it will get popped.

I don't remember anymore why I had to do this as a two-pass thing, making the first plan that explains the stop privileged.  That's an odd asymmetry, and it seems like going back to the "explains stop" loop after processing a previous "I explain the stop" plan will fix this issue.  But it will probably require a bunch of fiddling to get right again (and maybe discovering why I did it this way in the first place)...  So that should be a later resort.

Jim

**Comment by jeffreytan81** [comment]
> Why does the ThreadPlanSingleThreadTimeout have to claim to explain stops (like a random breakpoint hit) that on the face of it it's not responsible for

We are doing so to ensure `ThreadPlanSingleThreadTimeout` can stay in the leaf thread plan. And it comes from your suggestion `So ThreadPlanSingleThreadTimeout should always say it explains the stop` in this comment:
https://discourse.llvm.org/t/improve-single-thread-stepping/74599/17?u=jeffreytan81

**Comment by jimingham** [comment]
> > Why does the ThreadPlanSingleThreadTimeout have to claim to explain stops (like a random breakpoint hit) that on the face of it it's not responsible for
> 
> Just a quick comment to add the context for this question: we are doing so to ensure `ThreadPlanSingleThreadTimeout` can stay in the leaf thread plan. And it comes from your suggestion `So ThreadPlanSingleThreadTimeout should always say it explains the stop` in this comment: https://discourse.llvm.org/t/improve-single-thread-stepping/74599/17?u=jeffreytan81

So maybe we need to do something explicit here.  

The ThreadPlanSingleThreadTimeout's desired behavior, IIUC, is either it explains the stop or it should get discarded.  I can't think of any instance where it should stay on the stack when there's a stop that's not for the timeout.  If, for instance, you hit a "next range" breakpoint for a "step over" then we're going to remove the plan and have the thread plan push it again when you resume.

So instead of semi-lying about whether it actually explains stops it knows nothing about, can we add the concept of a plan that "has to be a leaf plan".  So if it doesn't explain a stop during ShouldStop, it will get popped right away.  So after we check "plan explains stop" and get `false` back, we check IsLeafPlan, and if that returns true, pop the plan and go on to the next plan.

The harder option is to either convert the ShouldStop to a single loop that re-enters if the plan that explained the stop returns True to MischiefManaged, or to figure out why the second loop is behaving asymmetrically to the first.  We ought to do the first at some point, but that's probably a lot more work.

**Comment by jeffreytan81** [comment]
@jimingham, it has been a while. Any other major changes you want to see? Thanks

File: lldb/include/lldb/Target/ThreadPlan.h [hunk 27]
```diff
@@ -395,6 +396,11 @@ class ThreadPlan : public std::enable_shared_from_this<ThreadPlan>,
 
   bool IsControllingPlan() { return m_is_controlling_plan; }
 
+  // Returns true if this plan is a leaf plan, meaning the plan will be popped
```
Comment (jimingham) [inline lldb/include/lldb/Target/ThreadPlan.h:399]: Leaf plans are only auto-popped if they don't explain the stop, right?  Might be worth mentioning that here.

File: lldb/source/Target/Thread.cpp [hunk 28]
```diff
@@ -813,12 +819,17 @@ bool Thread::ShouldStop(Event *event_ptr) {
...
-
   if (!current_plan->PlanExplainsStop(event_ptr)) {
     if (current_plan->TracerExplainsStop()) {
       done_processing_current_plan = true;
       should_stop = false;
     } else {
+      // Leaf plan that does not explain the stop should be popped.
+      // The plan should be push itself later again before resuming to stay
+      // as leaf.
+      if (current_plan->IsLeafPlan() && current_plan->MischiefManaged())
```
Comment (jimingham) [inline lldb/source/Target/Thread.cpp:830]: If the plan is a leaf plan and doesn't explain the stop, it should always be popped, shouldn't it?  Why do you include the MischiefManaged check here?

**Comment by jimingham** [comment]
Ack, sorry, I missed that you had updated this.  My notifications from GitHub are somewhat flakey...

I didn't understand why you are checking both "IsLeafPlan" and "MischiefManaged", why does MischiefManaged get to override IsLeafPlan?

File: lldb/source/Target/Thread.cpp (hunk 28 above)
Comment (jeffreytan81) [reply lldb/source/Target/Thread.cpp:830]: Thanks. I thought I had a reason to add this check 3 weeks ago but could not remember anymore. Running all the tests without the check pass so it might be a legacy change. Removed.

**Review by jimingham — APPROVED** [review]
This LGTM at this point.

**Comment by adrian-prantl** [comment]
@jimingham Should we revert the patch until this is fixed or are there more patches depending on this?

**Comment by jeffreytan81** [comment]
@jimingham, @adrian-prantl , thanks for letting me know. Surprised that I did not get failure email. 
Do you know how I can reproduce the failure?

**Comment by jimingham** [comment]
The console logs for the bots usually have the CMake  command used to configure the tree for the test.  The one I referred to above has:

'/usr/local/bin/cmake' '-G' 'Ninja' '/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-sanitized/llvm-project/llvm' '-DCMAKE_BUILD_TYPE=Release' '-DCMAKE_EXPORT_COMPILE_COMMANDS=ON' '-DLLVM_TARGETS_TO_BUILD=X86;ARM;AArch64' '-DCMAKE_INSTALL_PREFIX=/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-sanitized/lldb-install' '-DCMAKE_MAKE_PROGRAM=/usr/local/bin/ninja' '-DLLDB_TEST_USER_ARGS=--build-dir;/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-sanitized/lldb-build/lldb-test-build.noindex;-t;--env;TERM=vt100' '-DLLDB_ENFORCE_STRICT_TEST_REQUIREMENTS=On' '-DLLDB_ENABLE_PYTHON=On' '-DLLDB_ENABLE_LZMA=Off' '-DLIBCXX_HARDENING_MODE=none' '-DLLVM_ENABLE_ASSERTIONS:BOOL=TRUE' '-DLLVM_ENABLE_MODULES=Off' '-DLLVM_ENABLE_PROJECTS=clang;lld;lldb' '-DLLVM_LIT_ARGS=-v --time-tests --shuffle --xunit-xml-output=/Users/ec2-user/jenkins/workspace/llvm.org/lldb-cmake-sanitized/test/results.xml -v -j 3 --timeout 1200' '-DLLVM_VERSION_PATCH=99' '-DLLVM_ENABLE_RUNTIMES=libcxx;libcxxabi;libunwind' '-DLLVM_TARGETS_TO_BUILD=X86' '-DLLVM_USE_SANITIZER=Address;Undefined'

I think the only part of that that's relevant for you is `-DLLVM_USE_SANITIZER=Address;Undefined` - you should be able to just add that to your CMakery, then run the test.

**Comment by jeffreytan81** [comment]
I think I understand the lifetime issue that caused the ASAN failure. I will fix it.

**Comment by jeffreytan81** [comment]
Should fixed by https://github.com/llvm/llvm-project/pull/102208.

**Comment by labath** [comment]
The new test is still hanging occasionally (e.g. https://lab.llvm.org/buildbot/#/builders/162/builds/3718/steps/6/logs/stdio ). I don't know if it's a problem with the code being tested, or the test itself, but it looks like there's still some issues here.

**Comment by jeffreytan81** [comment]
@labath, unfortunately the failing log does not provide much information to troubleshoot. Is it possible to gather more information of the failure? Like is it only this `lldb-x86_64-debian` bot timeout or including other bots? And is it consistently failing or occasionally (race condition)? I tried to navigate the buildbot UI but couldn't make sense or find answers to these questions. 

We only have CentOS9 Linux machines and have run this PR clean build multiple times but couldn't reproduce the failure/hang.
 
Maybe we could try to disable the tests for Debian Linux (if it only happens in Debian bot)? Any suggestion is welcome.

**Comment by labath** [comment]
It's not just this bot. I also found failures on [aarch64](https://lab.llvm.org/buildbot/#/builders/59/builds/2999) and [arm](https://lab.llvm.org/buildbot/#/builders/18/builds/2079).  I don't see it on the mac bots, but I don't know how to look further into the history for those. The only way I can reproduce the problem is the same way that you can -- run the test many many times and try to catch it happening. For a relatively rare failure like this you may need to run it hundreds of times (and also put some background load on the system to make scheduling more nondeterministic) to catch it. I'm pretty sure CentOS is not immune to this bug.

**Comment by jeffreytan81** [comment]
Thanks for sharing. I have written a shell script to run `TestSingleThreadStepTimeout.py` in a loop infinitely until failing. Unfortunately, it has been running for 127 iterations (more than 7 hours) on my CentOS9 machine, haven't seen a single failure/timeout yet. I am not confident that I can reproduce with the linux machine I have.

Assuming it is only linux platforms, If this is blocking stuff, I am thinking of disabling the tests for Linux platform until a solid reproduce can be found to investigate. Let me know if you have any other thoughts.

**Comment by labath** [comment]
This is not specific to linux. I've now figured out how to scroll the green dragon output, and I've found [at](https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/9537/) [least](https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/9474/) [three](https://green.lab.llvm.org/job/llvm.org/view/LLDB/job/as-lldb-cmake/9457/) failures on mac as well.

I don't think disabling the test everywhere is a solution. If we can't get this resolved quickly (O(days)), we should revert the patch and continue the investigation offline. I've captured the [gdb packet log](https://paste.debian.net/hidden/1c7c754e/) from one of the failing runs, maybe that can get you started?

**Comment by jeffreytan81** [comment]
The GDB packet log is very useful (especially with a good run log at the top to compare). 

I read with @clayborg . The log indicates that "self.thread.StepOver(lldb.eOnlyThisThread)" sent a "$vCont;c:p1a24.1a24" to continue single thread at the end but lacking a "\x03" async interrupt packet (which can be found in the good log at the top). The async interrupt should be sent by `ThreadPlanSingleThreadTimeout::TimeoutThreadFunc()` timer thread after timeout. Further reading the code, we found that there is a potential race that timer thread checks the m_isAlive flag before it was initialized. 

Created https://github.com/llvm/llvm-project/pull/104195 to fix the race.
